### PR TITLE
Outputs ENR text representation in admin.nodeInfo RPC

### LINKED ISSUE #5622: Output ENR text representation in admin.nodeInfo RPC
See https://github.com/ethereum/EIPs/blob/master/EIPS/eip-778.md#text-encoding for the relevant part of ENR spec.

1. Adapt base64 encoding/decoding functions to support URL-safe encoding (basically use `-` instead of `+`, `_` instead of `/` and don't pad with `=`)
https://github.com/ethereum/aleth/blob/master/libdevcore/Base64.h

2. Add method to `ENR` class returning string with text form of ENR (serilize to RLP and base64-encode it)

3. Add `enr` field in the output of `admin_net_nodeInfo` https://github.com/ethereum/aleth/blob/6890d9a8548008419dd4bc054ba7908eac3a933f/libweb3jsonrpc/AdminNet.cpp#L27
You'll need to extend `dev::p2p::NodeInfo` struct for that and fill it in the method `Host::nodeInfo()` with `Host::enr()`.

### PR DESCRIPTION
**Pull request by twinstar26**
Fixes #5622
Old PR #5697

### CODE REVIEW

**Comment by twinstar26** [comment]
@gumb0  based on reviews in old PR #5697 

Everything seems to be working fine. Also can you double check changes in `libdevcore/Base64.cpp`.
If everything is fine I will go on to create a unit test as suggested.

File: libdevcore/Base64.cpp [hunk 1]
```diff
@@ -46,13 +46,8 @@ static inline byte find_base64_char_index(byte c)
 	else return 1 + find_base64_char_index('/');
 }
 
-string dev::toBase64(bytesConstRef _in)
+string toBase64Encoding(bytesConstRef _in, char const* base64_chars, bool _pad)
```
Comment (gumb0) [inline libdevcore/Base64.cpp]:
```suggestion
string toBase64Encoding(bytesConstRef _in, char const* _base64_chars, bool _pad)
```

File: libdevcore/Base64.cpp [hunk 2]
```diff
@@ -91,13 +86,33 @@ string dev::toBase64(bytesConstRef _in)
...
 	return ret;
 }
 
+string dev::toBase64(bytesConstRef _in)
+{
+	static const char base64_chars[] =
+		"ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+		"abcdefghijklmnopqrstuvwxyz"
+		"0123456789+/";
+	bool _pad = true;
```
Comment (gumb0) [inline libdevcore/Base64.cpp]:
```suggestion
	bool const pad = true;
```

File: libdevcore/Base64.cpp [hunk 3]
```diff
@@ -91,13 +86,33 @@ string dev::toBase64(bytesConstRef _in)
...
+	return toBase64Encoding(_in, base64_chars, _pad);
+}
+
+string dev::toBase64URLSafe(bytesConstRef _in)
+{
+	static const char base64_chars[] =
+		"ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+		"abcdefghijklmnopqrstuvwxyz"
+		"0123456789-_";
+	bool _pad = false;
```
Comment (gumb0) [inline libdevcore/Base64.cpp]:
```suggestion
	bool const pad = false;
```

File: libdevcore/Base64.cpp [hunk 4]
```diff
@@ -91,13 +86,33 @@ string dev::toBase64(bytesConstRef _in)
...
+		while (i++ < 3 && _pad)
 			ret += '=';
 	}
 
 	return ret;
 }
 
+string dev::toBase64(bytesConstRef _in)
+{
+	static const char base64_chars[] =
```
Comment (gumb0) [inline libdevcore/Base64.cpp]:
```suggestion
	static char const base64_chars[] =
```

File: libdevcore/Base64.cpp [hunk 5]
```diff
@@ -91,13 +86,33 @@ string dev::toBase64(bytesConstRef _in)
...
+		"ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+		"abcdefghijklmnopqrstuvwxyz"
+		"0123456789+/";
+	bool _pad = true;
+	return toBase64Encoding(_in, base64_chars, _pad);
+}
+
+string dev::toBase64URLSafe(bytesConstRef _in)
+{
+	static const char base64_chars[] =
```
Comment (gumb0) [inline libdevcore/Base64.cpp]:
```suggestion
	static char const base64_chars[] =
```

File: libp2p/ENR.cpp [hunk 6]
```diff
@@ -279,5 +280,12 @@ std::ostream& operator<<(std::ostream& _out, ENR const& _enr)
     return _out;
 }
 
+std::string ENR::textEncoding() const
+{
+    RLPStream s;
+    this->streamRLP(s);
```
Comment (gumb0) [inline libp2p/ENR.cpp]:
```suggestion
    streamRLP(s);
```

File: libp2p/ENR.cpp [hunk 7]
```diff
@@ -279,5 +280,12 @@ std::ostream& operator<<(std::ostream& _out, ENR const& _enr)
...
+    return toBase64URLSafe(std::string(s.out().begin(), s.out().end()));
```
Comment (gumb0) [inline libp2p/ENR.cpp]: I think something like 
```suggestion
    return toBase64URLSafe(ref(s.out()));
```
should work

File: libp2p/Host.h [hunk 8]
```diff
@@ -240,7 +241,14 @@ class Host: public Worker
     }
 
     /// Get the node information.
-    p2p::NodeInfo nodeInfo() const { return NodeInfo(id(), (networkConfig().publicIPAddress.empty() ? m_tcpPublic.address().to_string() : networkConfig().publicIPAddress), m_tcpPublic.port(), m_clientVersion); }
+    p2p::NodeInfo nodeInfo() const
+    { 
+        auto e = enr();
```
Comment (gumb0) [inline libp2p/Host.h]:
```suggestion
        auto const e = enr();
```

**Comment by gumb0** [comment]
For unit test add a check that `enr.textEncoding()` is `enr:-IS4QHCYrYZbAKWCBRlAy5zzaDZXJBGkcnh4MHcBFZntXNFrdvJjX04jRzjzCBOonrkTfj499SZuOh8R33Ls8RRcy5wBgmlkgnY0gmlwhH8AAAGJc2VjcDI1NmsxoQPKY0yuDUmstAHYpMa2_oxVtw0RW_QAdpzBQA8yWM0xOIN1ZHCCdl8` to this test https://github.com/ethereum/aleth/blob/b212789669dec65b342c5208cf4343632cc83b14/test/unittests/libp2p/ENRTest.cpp#L24
(this is a test vector from EIP spec)

**Comment by twinstar26** [comment]
> Ah, one more thing: you need to add a prefix `enr:` in function `ENR::textEncoding()`
> See the spec https://github.com/ethereum/EIPs/blob/master/EIPS/eip-778.md#text-encoding

@gumb0 
Thanks for the reviews!
So I was thinking of string concat of "enr:" and enrtext since `ENR::textEncoding()` returns string.

**Comment by twinstar26** [comment]
> For unit test add a check that `enr.textEncoding()` is `enr:-IS4QHCYrYZbAKWCBRlAy5zzaDZXJBGkcnh4MHcBFZntXNFrdvJjX04jRzjzCBOonrkTfj499SZuOh8R33Ls8RRcy5wBgmlkgnY0gmlwhH8AAAGJc2VjcDI1NmsxoQPKY0yuDUmstAHYpMa2_oxVtw0RW_QAdpzBQA8yWM0xOIN1ZHCCdl8` to this test
> 
> https://github.com/ethereum/aleth/blob/b212789669dec65b342c5208cf4343632cc83b14/test/unittests/libp2p/ENRTest.cpp#L24
> 
> (this is a test vector from EIP spec)

So after this do I have to create a unit-test explicitly and if so what is that (output of which function) I have to test.

**Comment by gumb0** [comment]
> So I was thinking of string concat of "enr:" and enrtext since ENR::textEncoding() returns string.

Yes, just return `"enr:" + toBase64URLSafe(...)` from `textEncoding()`

> So after this do I have to create a unit-test explicitly and if so what is that (output of which function) I have to test.

Let's add just this one check
```
EXPECT_EQ(enr.textEncoding(), "enr:-IS4QHCYrYZbAKWCBRlAy5zzaDZXJBGkcnh4MHcBFZntXNFrdvJjX04jRzjzCBOonrkTfj499SZuOh8R33Ls8RRcy5wBgmlkgnY0gmlwhH8AAAGJc2VjcDI1NmsxoQPKY0yuDUmstAHYpMa2_oxVtw0RW_QAdpzBQA8yWM0xOIN1ZHCCdl8")
``` 
to the end of the test I linked, I think that would be enough for this PR.

**Comment by gumb0** [comment]
Also please add an item to CHANGELOG.md

**Comment by twinstar26** [comment]
@gumb0 
This is merge ready. Please check and if not so please suggest changes.

File: CHANGELOG.md [hunk 9]
```diff
@@ -15,6 +15,7 @@
 - Added: [#5634](https://github.com/ethereum/aleth/pull/5634) Bootnodes for Rinkeby and Goerli.
 - Added: [#5640](https://github.com/ethereum/aleth/pull/5640) Istanbul support: EIP-1702 Generalized Account Versioning Scheme.
 - Added: [#5690](https://github.com/ethereum/aleth/issues/5690) Istanbul support: EIP-2028 transaction data gas cost reduction.
+- Added: [#5622](https://github.com/ethereum/aleth/issues/5622) Outputs ENR text representation in admin.nodeInfo RPC.
```
Comment (gumb0) [inline CHANGELOG.md]: Please link this to this PR instead of an issue

**Comment by chfast** [comment]
Merge?

**Pull request merged**